colossus/src/main/scala/colossus/protocols/http/Header.scala

```diff
@@ -62,37 +64,96 @@ object HttpHeaders {
   val CookieHeader      = "cookie"
   val SetCookie         = "set-cookie"
   val TransferEncoding  = "transfer-encoding"
+
+  def apply(hdrs: HttpHeader*) : HttpHeaders = new HttpHeaders(hdrs.toArray)
+}
+
+trait HttpHeader {
+  def key: String
+  def value: String
+  def encoded: Array[Byte]
+
+  override def equals(that: Any): Boolean = that match {
+    case that: HttpHeader => this.key == that.key && this.value == that.value
+    case other => false
+  }
+
+  override def hashCode = (key + value).hashCode
+
+  override def toString = s"($key,$value)"
+
+}
+
+//generally created when encoding http responses
+class EncodedHeader(val encoded: Array[Byte], keyLength: Int, valueStart: Int) extends HttpHeader {
+  lazy val key = new String(encoded.take(keyLength))
+  lazy val value = new String(encoded.drop(valueStart))
 }
 
-//TODO: support for pulling values as types other than String
-trait HttpHeaderUtils {
+//generally created when parsing http requests
+class DecodedHeader(val key: String, val value: String) extends HttpHeader {
+  lazy val encoded = (key + ": " + value).getBytes("UTF-8")
 
-  def headers : Seq[(String, String)]
+  def toEncodedHeader = new EncodedHeader(encoded, key.length, key.length + 2)
+}
+
+
+object HttpHeader {
+  def apply(key: String, value: String): HttpHeader = (new DecodedHeader(key, value)).toEncodedHeader
 
 
+  object Conversions {
+    implicit def stringTuple2Header(t: (String, String)): HttpHeader = HttpHeader(t._1, t._2)
+    implicit def seqStringTuple2Headers(t: Seq[(String, String)]): HttpHeaders = new HttpHeaders(t.map{stringTuple2Header}.toArray)
+  }
+
+}
+    
+class HttpHeaders(private val headers: Array[HttpHeader]) {
   def singleHeader(name: String): Option[String] = {
     val l = name.toLowerCase
-    headers.collectFirst{ case (`l`, v) => v}
+    headers.collectFirst{ case x if (x.key == l) => x.value }
   }
 
   def multiHeader(name: String): Seq[String] = {
```

Comment:
I actually don't think we need to repeat "header" in the name, but maybe the names `firstValue` and `allValues` might be better.